asf_public_discourse_home_decarbonisation/utils/text_processing_utils.py

```diff
@@ -0,0 +1,203 @@
+from gensim.parsing.preprocessing import STOPWORDS
+import re
+import string
+from nltk.stem.wordnet import WordNetLemmatizer
+from nltk import pos_tag
+from nltk.corpus import wordnet
+
+nltk.download("punkt")
+nltk.download("averaged_perceptron_tagger")
+nltk.download("wordnet")
+import pandas as pd
+from collections import Counter
+
+
+def remove_urls(text: str) -> str:
+    """
+    Removes URLs from text.
+    Args:
+        text (str): a string, tipically one or multiple sentences long
+    Returns:
+        str: text without URLs
+    """
+    pattern = r"http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\\(\\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+"
+    cleaned_text = re.sub(pattern, " ", text)
+    return cleaned_text
+
+
+def replace_punctuation_with_space(text: str) -> str:
+    """
+    Replaces punctuation with space.
+    Args:
+        text (str): a string, tipically one or multiple sentences long
+    Returns:
+        str: text wihtout punctuation
+    """
+    pattern = r"[{}]".format(re.escape(string.punctuation))
+    replaced_text = re.sub(pattern, " ", text)
+    replaced_text = re.sub(r"\s+", " ", replaced_text)  # Remove consecutive spaces
+    return replaced_text.strip()  # Remove leading/trailing spaces
+
+
+def remove_text_after_patterns(text: str) -> str:
+    """
+    Removes pattern of the form "xxx writes: ".
+
+    Args:
+        text (str): text to be cleaned
+
+    Returns:
+        str: cleaned text
+    """
+    # We use re.sub() to replace the pattern with an empty string
+    result = re.sub(r"\w+ wrote »", " ", text)
+    return result
+
+
+def identify_part_of_speech(pos_tag: str) -> str:
+    """
+    Convert the tag given by nltk.pos_tag to the tag used by wordnet.
+
+    Args:
+        pos_tag (str): NLTk part of speach tag
+
+    Returns:
+        str: WordNet part os speach tag
+    """
+    if pos_tag.startswith("N"):
+        return wordnet.NOUN
+    elif pos_tag.startswith("V"):
+        return wordnet.VERB
+    elif pos_tag.startswith("R"):
+        return wordnet.ADV
+    elif pos_tag.startswith("J"):
+        return wordnet.ADJ
+    else:
+        return wordnet.NOUN  # Default to NOUN if the part of speech is not recognized
+
+
+def lemmatize_sentence(tokens: list) -> dict:
+    """
+    Lemmatizes tokens using WordNetLemmatizer.
+
+    Args:
+        tokens (list): a list of tokens
+
+    Returns:
+        dict: a dictionary mapping each token to the corresponding lemmatized token
+    """
+    lemmatizer = WordNetLemmatizer()
+    pos_tags = pos_tag(tokens)
+    lemmatized_words_dict = {
+        word: lemmatizer.lemmatize(word, identify_part_of_speech(pos_tag))
+        for word, pos_tag in pos_tags
+    }
+    return lemmatized_words_dict
+
+
+def process_text(text: str) -> str:
+    """
+    Preprocesses text by:
+        - replacing &amp with "and"
+        - removing URLs
+        - puts text to lower case
+        - replaces punctuation with space
+        - applies lemmatisation
+
```

Comment:
Do we think we could add a function here to remove words of <=2?

Reply:
Thinking about this further, we would get rid of units like dB and £, so scratch that!

Reply:
hasn't been a problem so far to have them :) was it a problem for you with BuildHub?

Reply:
removing stop words would handle any short unnecessary words anyway